### PR TITLE
executor: error processing refactor for load data

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve? 

there are two different modifications(maybe they should not be considered together?)

1. remove old logic in load data error processing,  `commitWork` function will commit or rollback txn. Former implementation will make query processing panic since commit fail and `txn` is invalid.
Ealier  tidb version still has this problem(2.1 3.0), you will get `lost connection` error if doing sth like ddl between `load data`

2. `doCommit` function defer will make txn `changeToInvalid`, so if `commit` failed and `stmtRollBack` is called, `txnState` has nil `transaction` field,  which will cause `ConfirmAssertions`  operation report panic error "invalid memory nil"

some unit-test or tidb-test will be added for this situation

```
enable commit fail point in `commitOneTask` function
do load data query

client expected like:
ERROR 1105 (HY000): Information schema is changed. [try again later]
not `lost connection` and no panic error log
```

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
```
enable commit fail point in `commitOneTask` function
do load data query

client expected like:
ERROR 1105 (HY000): Information schema is changed. [try again later]
not `lost connection` and no panic error log
```

Code changes

Side effects

Related changes

Release note

### CODE REVIEW

File: server/conn.go [hunk 1]
```diff
@@ -1153,23 +1153,7 @@ func (cc *clientConn) handleLoadData(ctx context.Context, loadDataInfo *executor
...
-		if txn != nil && txn.Valid() {
-			if err != nil {
-				if err1 := txn.Rollback(); err1 != nil {
-					logutil.Logger(ctx).Error("load data rollback failed", zap.Error(err1))
-				}
-				return err
-			}
-			return cc.ctx.CommitTxn(sessionctx.SetCommitCtx(ctx, loadDataInfo.Ctx))
-		}
-	}
```
Comment (lysu) [inline server/conn.go]: remove this snippet will remove function added by https://github.com/pingcap/tidb/pull/9444?

maybe some reason we need `txn.Rollback` here if `txn.Valid` and have `err`, PTAL @crazycs520

Comment (cfzjywxk) [reply server/conn.go]: former implementation will leave one batch/txn undecided out for loop, so the main logic need to  process the last txn

File: session/txn.go [hunk 2]
```diff
@@ -461,7 +461,9 @@ func (s *session) StmtCommit() error {
 // StmtRollback implements the sessionctx.Context interface.
 func (s *session) StmtRollback() {
```
Comment (cfzjywxk) [inline session/txn.go]: @tiancaiamao PTAL  if we can change `StmtRollback` like this

File: server/server_test.go [hunk 3]
```diff
@@ -684,6 +685,59 @@ func runTestLoadData(c *C, server *Server) {
...
+
+		// fail error processing test
+		dbt.Assert(failpoint.Enable("github.com/pingcap/tidb/executor/commitOneTaskErr", "return"), IsNil)
+		_, err1 = dbt.db.Exec(`load data local infile '/tmp/load_data_test.csv' into table pn FIELDS TERMINATED BY ','`)
+		mysqlErr, ok := err1.(*mysql.MySQLError)
+		dbt.Assert(ok, IsTrue)
+		dbt.Assert(mysqlErr.Message, Equals, "mock commit one task error")
+		dbt.Assert(failpoint.Disable("github.com/pingcap/tidb/executor/commitOneTaskErr"), IsNil)
+
+		//
```
Comment (lonng) [inline server/server_test.go]:
```suggestion
```

File: session/txn.go (hunk 2 above)
Comment (tiancaiamao) [reply session/txn.go]: That's ok.
B.T.W, I'm thinking of removing the assertion proto which is introduced to add an additional check for write consistency, but I find it's hard to achieve that goal.

**Comment by tiancaiamao** [comment]
LGTM @crazycs520

File: server/conn.go (hunk 1 above)
Comment (jackysp) [reply server/conn.go]: PTAL @crazycs520

Comment (cfzjywxk) [reply server/conn.go]: @jackysp  PTAL

**Pull request merged**